rules/scripts/bitstreams_workspace.py

```diff
@@ -0,0 +1,301 @@
+parser.add_argument('--refresh', default=False, action=argparse.BooleanOptionalAction,
+                    help='Force a refresh')
+parser.add_argument('--refresh-time', default=300, type=int,
+                    help='How often to check for new bitstreams')
+parser.add_argument('--repo', default='', help="Location of the source git repo")
+parser.add_argument('bitstream', default='latest', nargs='?',
+                    help='Bitstream to retrieve: "latest" or git commit identifier')
+
+
+class BitstreamCache(object):
+
+    def __init__(self, bucket_url, cachedir, latest_update, offline=False):
+        """Initialize the Bitstream Cache Manager."""
+        if bucket_url[-1] != '/':
+            bucket_url += '/'
+        self.bucket_url = bucket_url
+        cachedir = os.path.expanduser(cachedir)
+        self.cachedir = os.path.join(cachedir, 'cache')
+
+        latest_update = os.path.join(cachedir, os.path.expanduser(latest_update))
+        self.latest_update = latest_update
+        self.offline = offline
+        self.http = urllib3.PoolManager()
+        self.available = {}
+
+    def InitRepository(self):
+        """Create the cache directory and symlink it into the bazel repository dir."""
+        os.makedirs(self.cachedir, exist_ok=True)
+        os.symlink(self.cachedir, 'cache')
+
+    def Touch(self, key):
+        """Set the latest known bitstream.
+
+        Args:
+            key: str; The git hash of the latest bitstream.
+        """
+        with open(self.latest_update, 'w') as f:
+            f.write(key)
+
+    def NeedRefresh(self, interval):
+        """Determine if the cache needs a refresh.
+
+        Args:
+            interval: int; Desired interval between refresh.
+        Returns:
+            bool: whether a refresh is needed.
+        """
+        try:
+            st = os.stat(self.latest_update)
+            return time.time() - st.st_mtime > interval
+        except FileNotFoundError:
+            return True
+
+    def Get(self, file):
+        """Perform an HTTP GET from the GCP bitstream bucket.
+
+        Args:
+            file: Filename in the bucket to retrieve.
+        Returns:
+            bytes
+        """
+        response = self.http.request('GET', self.bucket_url + file)
+        return response.data
+
+    def GetBitstreamsAvailable(self, refresh):
+        """Inventory which bitstreams are available.
+
+        Args:
+            refresh: bool; whether to refresh from the network.
+        """
+        if not refresh:
+            for (_, dirnames, _) in os.walk('cache'):
+                for d in dirnames:
+                    self.available[d] = 'local'
+            try:
+                with open(self.latest_update) as f:
+                    self.available['latest'] = f.read()
+            except FileNotFoundError:
+                pass
+            return
+        document = self.Get('').decode('utf-8')
+        et = xml.etree.ElementTree.fromstring(document)
```

Comment:
Is there a reason we're opting to parse the XML instead of using an existing Python api for gcp (e.g. https://cloud.google.com/storage/docs/listing-objects#storage-list-objects-python)

Reply:
The python API is huge and workspace rules have limitations on what they can depend on.  I'd considered downloading `gsutil` as a dependency, but its ~100MiB.